spec/codecs/netflow_spec.rb

```diff
@@ -1 +1,208 @@
 require "logstash/devutils/rspec/spec_helper"
+require "logstash/codecs/netflow"
+
+describe LogStash::Codecs::Netflow do
+  subject do
+    LogStash::Codecs::Netflow.new.tap do |codec|
+      expect{codec.register}.not_to raise_error
+    end
+  end
+
+  let(:decode) do
+    [].tap do |events|
+      subject.decode(data){|event| events << event}
+    end
+  end
+
+  context "Netflow 5" do
+    let(:data) do
+      # this netflow raw data was produced with softflowd and captured with netcat
+      # softflowd -D -i eth0 -v 5 -t maxlife=1 -n 127.0.01:8765
+      # nc -k -4 -u -l 127.0.0.1 8765 > netflow5.dat
+      IO.read(File.join(File.dirname(__FILE__), "netflow5.dat"), :mode => "rb")
+    end
+
+    let(:json_events) do
+      events = []
+      events << <<-END
+        {
+          "@timestamp": "2015-05-02T18:38:08.280Z",
+          "netflow": {
+            "version": 5,
+            "flow_seq_num": 0,
+            "engine_type": 0,
+            "engine_id": 0,
+            "sampling_algorithm": 0,
+            "sampling_interval": 0,
+            "flow_records": 2,
+            "ipv4_src_addr": "10.0.2.2",
+            "ipv4_dst_addr": "10.0.2.15",
+            "ipv4_next_hop": "0.0.0.0",
+            "input_snmp": 0,
+            "output_snmp": 0,
+            "in_pkts": 5,
+            "in_bytes": 230,
+            "first_switched": "2015-06-21T11:40:52.280Z",
+            "last_switched": "2015-05-02T18:38:08.279Z",
+            "l4_src_port": 54435,
+            "l4_dst_port": 22,
+            "tcp_flags": 16,
+            "protocol": 6,
+            "src_tos": 0,
+            "src_as": 0,
+            "dst_as": 0,
+            "src_mask": 0,
+            "dst_mask": 0
+          },
+          "@version": "1"
+        }
+      END
+
+      events << <<-END
+        {
+          "@timestamp": "2015-05-02T18:38:08.280Z",
+          "netflow": {
+            "version": 5,
+            "flow_seq_num": 0,
+            "engine_type": 0,
+            "engine_id": 0,
+            "sampling_algorithm": 0,
+            "sampling_interval": 0,
+            "flow_records": 2,
+            "ipv4_src_addr": "10.0.2.15",
+            "ipv4_dst_addr": "10.0.2.2",
+            "ipv4_next_hop": "0.0.0.0",
+            "input_snmp": 0,
+            "output_snmp": 0,
+            "in_pkts": 4,
+            "in_bytes": 304,
+            "first_switched": "2015-06-21T11:40:52.280Z",
+            "last_switched": "2015-05-02T18:38:08.279Z",
+            "l4_src_port": 22,
+            "l4_dst_port": 54435,
+            "tcp_flags": 24,
+            "protocol": 6,
+            "src_tos": 0,
+            "src_as": 0,
+            "dst_as": 0,
+            "src_mask": 0,
+            "dst_mask": 0
+          },
+          "@version": "1"
+        }
+      END
+
+      events.map{|event| event.gsub(/\s+/, "")}
+    end
+
+    it "should decode raw data" do
```

Comment:
@purbon I am sure I can learn a lot from your test/spec skills but this comment is not constructive and just not true.

I do not think you took the time to understand the test purposes. In the absolute, these tests may not be the best tests a codec could deserve but it is way better than **no test** as it was before. These tests are the best I could come up with in my work on this bug on a Saturday afternoon. 

The purposes are:

1- Make sure the serialization bug this PR fixes is covered by the test. This is done in the `"should serialize to json"` for both `"Netflow 5"` and `"Netflow 9"` contexts. I do not see how much less we could check for this. This test is checking a single thing: will it serialize without crashing and is the resulting json valid. Now, notice I said "a single thing" but I listed 2, "without crashing" and "resulting json valid", this is because here "without crashing" is implicit, like in almost all tests we write, that the method we call is "without crashing", we don't need to explicitly check for rogue exceptions for every method we test. 

2- Since there was **no test** for this codec, I decided to play the "reverse engineering" test game, by feeding some raw data that I controlled, checking the output events, verify this output is correct from the environment parameters I controlled (IP addresses, ports, protocols) and use that output as a basis for a **first** decoding assertion. SO the `"should decode raw data"` test actually simply tests that decoded raw Netflow data actually generates some events with some expected fields. In other words, this is a *rough* **"does it decode or not"** type of test. This test will have the benefit of failing if there are refactoring or modification mistakes made in the codec in the future.

I am not sure what your *"you're checking for so many things in just one test"* means, I am actually just checking one thing: does it decode or not, and for this I am checking a few fields.  Did you suggest every assertion should have been in a single `it "should decode this fields ..."` style? I don't think so. This would have been way too verbose, and at this point, all I can do is say that if any of these field is not decoded properly, then there is something fishy in the decoding as a whole. 

If I was the original author and had better knowledge the Netflow protocol/format better, chances my tests would have been a lot more precise, but this is not the case. 

So, all in all, **I do not agree** these two test examples are wrong and **I am very satisfied** with what I came up with.